lib/languageFeatures/findReferencesHandler.ts

```diff
@@ -1,78 +1,119 @@
+        .filter(subprogram => subprogram.lexerToken.getLText() == definition.lexerToken?.getLText()).forEach(def => definitions.add(def));
+      if (definition.parent instanceof OPackage) {
+        definition.parent.correspondingPackageBodies.flatMap(packageBodies => packageBodies.subprograms
+          .filter(subprogram => subprogram.lexerToken.getLText() == definition.lexerToken?.getLText())).forEach(def => definitions.add(def));
+      }
+      if (definition.parent instanceof OPackageBody) {
+        ((definition.parent as OPackageBody).correspondingPackage?.subprograms
+          .filter(subprogram => subprogram.lexerToken.getLText() == definition.lexerToken?.getLText()) ?? []).forEach(def => definitions.add(def));
+      }
+
+
+    }
   }
-  // debugger;
-  if (candidate instanceof OReference && candidate.definitions) {
-    return candidate.definitions.concat(candidate.definitions.flatMap(c => {
-      return implementsIHasReference(c) ? c.referenceLinks : [];
-    }));
+  const definitionsNew = [...definitions].map(definition => {
+    if (definition instanceof OPackageBody && definition.correspondingPackage) {
+      return definition.correspondingPackage;
+    }
+    return definition;
+
+  });
+  const tokens: OLexerToken[] = [];
+  for (const definition of definitionsNew) {
+    if (implementsIHasReference(definition)) {
+      if (definition.lexerToken) {
+        tokens.push(definition.lexerToken);
+      }
+      tokens.push(...definition.referenceLinks.map(ref => ref.referenceToken).filter(token => token.getLText() === definition.lexerToken?.getLText()));
+      if (definition instanceof OEntity) {
+        tokens.push(...definition.correspondingArchitectures.map(arch => arch.entityName));
+        tokens.push(...definition.referenceLinks.flatMap(link => link instanceof OInstantiation ? link.componentName : []));
+      }
+      if (definition instanceof OPort) {
+        tokens.push(...definition.parent.referenceLinks
+          .flatMap(link => link instanceof OInstantiation ? link.portAssociationList?.children.flatMap(child => {
+            return child.formalPart.filter(formal => formal.referenceToken.getLText() === definition.lexerToken.getLText()).map(formal => formal.referenceToken);
+          }) ?? [] : []));
+      }
+      if (definition instanceof OPackage) {
+        for (const correspondingPackageBody of definition.correspondingPackageBodies) {
+          tokens.push(correspondingPackageBody.lexerToken);
+          if (correspondingPackageBody.endingLexerToken) {
+            tokens.push(correspondingPackageBody.endingLexerToken);
+          }
+        }
+      }
+    }
+    if (implementsIHasEndingLexerToken(definition)) {
+      tokens.push((definition as IHasEndingLexerToken).endingLexerToken as OLexerToken);
+    }
+
   }
-  if (implementsIHasReference(candidate)) {
-    return ([candidate] as ObjectBase[]).concat(candidate.referenceLinks ?? []);
+  const map = new Map<string, OLexerToken>();
```

Comment:
```suggestion
  // make sure to only return one reference per range
  const map = new Map<string, OLexerToken>();
```
This is the goal of this, isn't it? Seems a bit hacky but I don't have a nicer approach atm

Reply:
Yes. If the references are doubled then vscode throws an error